neuralmonkey/attention/scaled_dot_product.py

```diff
@@ -23,6 +24,143 @@
+    # weights_all shape: batch, head, time(q), time(k)
+    weights_all = weights_4d * tf.expand_dims(tf.expand_dims(mask, 1), 1)
+
+    # normalization along time(k)
+    # norm shape: batch, head, time(q), 1
+    norm = tf.reduce_sum(weights_all, 3, keep_dims=True) + 1e-8
+    return weights_all / norm
+
+
+def mask_future(energies: tf.Tensor) -> tf.Tensor:
+    """Mask energies of keys to the right of the query."""
+    triangular_mask = tf.matrix_band_part(tf.ones_like(energies), -1, 0)
+    mask_area = tf.equal(triangular_mask, 1)
+    masked_value = tf.fill(tf.shape(energies), -np.inf)
+    return tf.where(mask_area, energies, masked_value)
+
+
+def attention(
+        queries: tf.Tensor,
+        keys: tf.Tensor,
+        values: tf.Tensor,
+        keys_mask: tf.Tensor,
+        num_heads: int,
+        dropout_callback: Callable[[tf.Tensor], tf.Tensor],
+        masked: bool = False) -> tf.Tensor:
+
+    if num_heads <= 0:
+        raise ValueError("Number of heads must be greater than zero.")
+
+    # Input shapes:
+    # queries:  batch, time(q), k_channels
+    # keys:   batch, time(k), k_channels
+    # values: batch, time(k), v_channels
+    #
+    # Output shapes:
+    # context: batch, time(q), v_channels
+    # weights: batch, time(q), time(k)
+
+    queries_dim = queries.get_shape()[-1].value
+    keys_dim = keys.get_shape()[-1].value
+
+    # Query and keys should match in the last dimension
+    if queries_dim != keys_dim:
+        raise ValueError("Queries and keys do not match in the last dimension."
+                         "Query: {}, Key: {}".format(queries_dim, keys_dim))
+
+    # Last dimension must be divisible by num_heads
+    if queries_dim % num_heads != 0:
+        raise ValueError(
+            "Last dimension of the query ({}) should be divisible by the "
+            "number of heads ({})".format(queries_dim, num_heads))
+
+    head_dim = int(queries_dim / num_heads)
+
+    # For multi-head attention, queries, keys and values are linearly projected
+    if num_heads > 1:
+        queries = tf.layers.dense(queries, queries_dim, name="query_proj")
+        keys = tf.layers.dense(keys, queries_dim, name="keys_proj")
+        values = tf.layers.dense(values, queries_dim, name="vals_proj")
+
+    # Scale first:
+    queries_scaled = queries / math.sqrt(head_dim)
+
+    # Reshape the k_channels dimension to the number of heads
+    queries = split_for_heads(queries_scaled, num_heads, head_dim)
+    keys = split_for_heads(keys, num_heads, head_dim)
+    values = split_for_heads(values, num_heads, head_dim)
+
+    # For dot-product, we use matrix multiplication
+    # shape: batch, head, time(q), time(k) (k_channels is the matmul axis)
+    energies = tf.matmul(queries, keys, transpose_b=True)
+
+    # To protect the attention from looking ahead of time, we must
+    # replace the energies of future keys with negative infinity
+    # We use lower triangular matrix and basic tf where tricks
```

Comment:
Ten posledni radek by snad i sel smaznout. Z te mask_future metody je to jasne.

Reply:
Jo, zvlášť až se vyřeší jindrův review u tý funkce `mask_future`